text/3697-declarative-attribute-macros.md

```diff
@@ -0,0 +1,127 @@
+- Feature Name: `declarative_attribute_macros`
+- Start Date: 2024-09-20
+- RFC PR: [rust-lang/rfcs#3697](https://github.com/rust-lang/rfcs/pull/3697)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Support defining `macro_rules!` macros that work as attribute macros.
+
+# Motivation
+[motivation]: #motivation
+
+Many crates provide attribute macros. Today, this requires defining proc
+macros, in a separate crate, typically with several additional dependencies
+adding substantial compilation time, and typically guarded by a feature that
+users need to remember to enable.
+
+However, many common cases of attribute macros don't require any more power
+than an ordinary `macro_rules!` macro. Supporting these common cases would
+allow many crates to avoid defining proc macros, reduce dependencies and
+compilation time, and provide these macros unconditionally without requiring
+the user to enable a feature.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a `macro_rules!` macro, you can prefix some of the macro's rules
+with `attr(...)` to allow using the macro as an attribute. The
+arguments to the attribute, if any, are parsed by the *MacroMatcher* in the
+first set of parentheses; the second *MacroMatcher* parses the entire construct
+the attribute was applied to. The resulting macro will work anywhere an
+attribute currently works.
+
+```rust
+macro_rules! main {
+    attr() ($func:item) => { make_async_main!($func) };
+    attr(threads = $threads:literal) ($func:item) => { make_async_main!($threads, $func) };
+}
+
+#[main]
+async fn main() { ... }
+
+#[main(threads = 42)]
+async fn main() { ... }
+```
+
+Attribute macros defined using `macro_rules!` follow the same scoping rules as
+any other macro, and may be invoked by any path that resolves to them.
+
+An attribute macro must not require itself for resolution, either directly or
+indirectly (e.g. applied to a containing module or item).
+
+Note that a single macro can have both `attr` and non-`attr` rules. Attribute
+invocations can only match the `attr` rules, and non-attribute invocations can
+only match the non-`attr` rules.
```

Comment:
The motivation for this is to support using the same name for both attribute macros and functionlike macros, iiuc. However, I think the current proposal would be pretty confusing - you more or less need to read all patterns and mentally separate which can be used as an attribute and which can be used in function form. Also diagnostics could get pretty weird - how would we know if a pattern forgot an `attr()` or if it just happened to match the wrong thing?

I think we should use `#[macro_attribute]`[^1] instead and then, if the motivation is strong enough, just allow defining both a `macro_attribute` and a plain (i.e. functionlike) MBE macro with the same name (possibly also a `#[macro_derive]` if that is helpful - I expect probably less so). Readability would be better because separate code blocks handle separate concerns, teachability improves because this is more similar to RFC3698 `#[macro_derive]`, and diagnostics would probably be more straightforward.

[^1]: `macro_attribute` mirrors `macro_derive` from [RFC3698](https://github.com/rust-lang/rfcs/pull/3698), but maybe `derive_macro`/`attribute_macro` would be better?